### PR TITLE
[mlir] Handle simple commutative cases in CSE.

### PR DESCRIPTION
**Pull request by jpienaar**
Tried to keep this simple while handling obvious CSE instances. For more complicated cases the expectation is still that the sorting pass would run before. While simple, this case did turn up in a real deployed instance where it had a large e2e impact. This can of course be refined.

### CODE REVIEW

File: mlir/lib/IR/OperationSupport.cpp [hunk 1]
```diff
@@ -696,7 +705,9 @@ llvm::hash_code OperationEquivalence::computeHash(
     Region *lhs, Region *rhs,
     function_ref<LogicalResult(Value, Value)> checkEquivalent,
     function_ref<void(Value, Value)> markEquivalent,
-    OperationEquivalence::Flags flags) {
+    OperationEquivalence::Flags flags,
+    function_ref<LogicalResult(ValueRange, ValueRange)>
+        checkCommutativeEquivalent) {
```
Comment (matthias-springer) [inline mlir/lib/IR/OperationSupport.cpp:710]: `checkCommutativeEquivalent` is not used in this function.

File: mlir/lib/IR/OperationSupport.cpp [hunk 2]
```diff
@@ -798,15 +844,24 @@ OperationEquivalence::isRegionEquivalentTo(Region *lhs, Region *rhs,
...
-    if (curArg == otherArg)
-      continue;
-    if (curArg.getType() != otherArg.getType())
-      return false;
-    if (failed(checkEquivalent(curArg, otherArg)))
+  if (checkCommutativeEquivalent &&
+      lhs->hasTrait<mlir::OpTrait::IsCommutative>()) {
+    auto lhsRange = lhs->getOperands();
+    auto rhsRange = rhs->getOperands();
+    if (failed(checkCommutativeEquivalent(lhsRange, rhsRange)))
```
Comment (matthias-springer) [inline mlir/lib/IR/OperationSupport.cpp:851]: Do we actually need `checkCommutativeEquivalent`? Can we do something like this if the op has the `IsCommutative` trait?

```c++
SmallVector<Value> rhsOperands(rhs->getOperands().begin(), rhs->getOperands().end());
for (Value lhsOperand : lhs->getOperands()) {
  auto it = llvm::find_if(rhsOperands, [&](Value rhsOperand) {
      return succeeded(checkEquivalent(lhsOperand, rhsOperand));
  });
  if (it == rhsOperands.end()) {
    // Could not find equivalent operand.
    return false;
  }
  rhsOperands.erase(it);
}
assert(rhsOperands.empty());
```

File: mlir/lib/IR/OperationSupport.cpp [hunk 3]
```diff
@@ -683,8 +683,17 @@ llvm::hash_code OperationEquivalence::computeHash(
     hash = llvm::hash_combine(hash, op->getLoc());
 
   //   - Operands
-  for (Value operand : op->getOperands())
-    hash = llvm::hash_combine(hash, hashOperands(operand));
+  if (op->hasTrait<mlir::OpTrait::IsCommutative>() &&
+      op->getNumOperands() > 0) {
+    // If commutative, don't hash the operands as hash is not order independent
+    // and even if it were would not be sufficient for CSE usage.
+    // FIXME: This has the effect of resulting in more hash collisions
```
[the remark this answers is not in the document]
Comment (jpienaar) [reply mlir/lib/IR/OperationSupport.cpp:690]: We could make it commutative here yes, but we actually need to hash based on the equivalence class additionally here to ensure they end with same hash before we compare. But we don't know those when we just hash.

File: mlir/lib/IR/OperationSupport.cpp (hunk 2 above)
Comment (jpienaar) [reply mlir/lib/IR/OperationSupport.cpp:851]: We could, but this is O(n^2) vs O(n log n). Now, this should be replaced by the explicit commutative sort pass in pipeline so that here it should suffice to just do the linear scan. The default param means its easy to drop again at that point without needing to update.

File: mlir/lib/IR/OperationSupport.cpp (hunk 1 above)
Comment (jpienaar) [reply mlir/lib/IR/OperationSupport.cpp:710]: Good catch, forgot to pass it in below.

File: mlir/include/mlir/IR/OperationSupport.h [hunk 4]
```diff
@@ -1271,7 +1271,9 @@ struct OperationEquivalence {
   isEquivalentTo(Operation *lhs, Operation *rhs,
                  function_ref<LogicalResult(Value, Value)> checkEquivalent,
                  function_ref<void(Value, Value)> markEquivalent = nullptr,
-                 Flags flags = Flags::None);
+                 Flags flags = Flags::None,
+                 function_ref<LogicalResult(ValueRange, ValueRange)>
+                     checkCommutativeEquivalent = nullptr);
```
Comment (joker-eph) [inline mlir/include/mlir/IR/OperationSupport.h:1276]: This should be documented.

I don't quite get why we need injection here by the way?

File: mlir/lib/IR/OperationSupport.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (joker-eph) [reply mlir/lib/IR/OperationSupport.cpp:690]: If the only "based on the equivalence" aspect is the commutative, then @matthias-springer seems right to me?
Are you anticipating other criteria?

File: mlir/lib/IR/OperationSupport.cpp [hunk 5]
```diff
@@ -751,6 +763,36 @@ struct ValueEquivalenceCache {
...
+
+    // Handle where operands in order are equivalent.
+    auto lhsIt = lhsRange.begin();
+    auto rhsIt = rhsRange.begin();
+    for (; lhsIt != lhsRange.end(); ++lhsIt, ++rhsIt) {
+      if (failed(checkEquivalent(*lhsIt, *rhsIt)))
+        break;
+    }
+    if (lhsIt == lhsRange.end())
+      return success();
```
Comment (joker-eph) [inline mlir/lib/IR/OperationSupport.cpp:780]: if (!llvm::all_of(llvm::zip(lhsRange, rhsRange), checkEquivalent) return success();`

Edit: oh you're trying to optimize to only sort a subrange?

File: mlir/lib/IR/OperationSupport.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpienaar) [reply mlir/lib/IR/OperationSupport.cpp:690]: AFAICS CSE ends up bucketing based on the hash of values and only those in same bucket are considered for comparison. Now that doesn't affect the current change as I'm just looking at reordering of values without considering equivalence below (except for the "simple" case, but that is equivalent to existing state).

File: mlir/include/mlir/IR/OperationSupport.h (hunk 4 above)
Comment (jpienaar) [reply mlir/include/mlir/IR/OperationSupport.h:1276]: As this is a static function and we need to be able to inject querying the equivalence classes. Else we have to do N^2 as Matthias suggested.

File: mlir/lib/IR/OperationSupport.cpp (hunk 5 above)
Comment (jpienaar) [reply mlir/lib/IR/OperationSupport.cpp:780]: Well, more I'm trying to avoid sorting. So once commutative is handled as presort somewhere, it would just be the above/commutative would not be special here. And the sorted side I'm not considering equivalence classes there, its pure Value comparison rather than querying `equivalentValues` map too. That's as mentioned in PR description to keep this rather minimal.

File: mlir/lib/IR/OperationSupport.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpienaar) [reply mlir/lib/IR/OperationSupport.cpp]: (this was sufficient for my test case, not the best symmetric merge, but I think good enough)

File: mlir/include/mlir/IR/OperationSupport.h (hunk 4 above)
Comment (joker-eph) [reply mlir/include/mlir/IR/OperationSupport.h:1279]: I'm not sure that answers my question, let me rephrase: why isn't this just a Flag?

Comment (jpienaar) [reply mlir/include/mlir/IR/OperationSupport.h:1279]: The equivalence classes are not exposed, so only accessible via the functors passed in. If a flag, then one could only access them via checkEquivalent still. Now, an option is to just not care about equivalence for commutative case except for pairwise part.

Comment (joker-eph) [reply mlir/include/mlir/IR/OperationSupport.h:1279]: OK, LG, but you're still missing the doc :)
